distributed/shuffle/shuffle-design.md

```diff
@@ -0,0 +1,139 @@
+- Hand off dask-managed data to be processed by the extension
+- Bring data produced out-of-band back into dask
+- Clean up when keys depending on the shuffle are cancelled
+
+The graph also has a secondary synchronization benefit, letting us bypass some difficult distributed problems (exactly-once initialization and cleanup tasks, determining when all peer-to-peer transfers are complete) by leaning on the scheduler.
+
+![diagram of graph](graph.png)
+
+### `ShuffleExtension`
+
+Problems to solve:
+* Holding per-worker out-of-band state for an in-progress shuffle
+* Adding new handlers in an organized way for workers to transfer shuffle data
+* Doing the above cleanly with multiple concurrent shuffles (`merge`)
+* Coordinating multiple concurrent shuffles which may need to share limited resources (memory, threads, etc.)
+* Getting metrics back to the scheduler/dashboard, like managed memory & bytes spilled to disk (eventually)
+
+The `ShuffleExtension` will be built into distributed and added to workers automatically (like the `PubSubWorkerExtension`). It'll add a route to the worker; something like:
+
+```python
+def shuffle_receive(comm, shuffle_id: str, data: DataFrame) -> None:
+    """
+    Receive an incoming shard of data from a peer worker.
+    Using an unknown ``shuffle_id`` will first initialize any state needed for that new shuffle.
+    """
+    ...
+```
+
+The `ShuffleExtension` will hold all per-shuffle state and buffers specific to that worker. For example, things like (will vary depending on the stage of the implementation):
+- a buffer of received shards
+- a buffer of outgoing shards (waiting to accumulate enough to be worth sending to a worker)
+- a datastore that transparently spills received shards to disk
+- locks or synchronization mechanisms
+
+The `ShuffleExtension` will also hold any global state shared by all shuffles, such as worker threads or coroutines.
+
+Most of the implementation-specific logic will happen in the `ShuffleExtension`. As we improve performance, add concurrency, etc., this code will change the most (though the interface will likely not).
+
+The `transfer` tasks will pass their input partitions into the `ShuffleExtension`, blocking the task until the extension is ready to receive another input. Internally, the extension will do whatever it needs to do to transfer the data, using worker comms to call the `shuffle_receive` RPC on peer workers. Simultaneously, it'll handle any incoming data from other workers.
+
+### Retries and cancellation
+
+Problems to solve:
+* Causing all tasks in the shuffle to rerun when a worker leaves
+* Cleaning up out-of-band state when a user cancels a shuffle, or it errors
+
+Because most the tasks in the shuffle graph are impure and run for their side effects, restarting an in-progress shuffle requires rerunning _every_ task involved, even ones that appear to have successfully transitioned to `memory` and whose "results" are stored on non-yet-dead workers.
```

Comment:
> The reason why we'd need to rerun all tasks is because of the hashing, isn't it?

Yes. If the number or order of workers has changed, then the hashing will make the same data go to a different worker the second time around. Your hypothetical is correct.